src/Internal/Payment/PaymentRequest.php

```diff
@@ -89,4 +106,42 @@ public function get_payment_method_id(): ?string {
 			? sanitize_text_field( wp_unslash( ( $this->request['payment_method_id'] ) ) )
 			: null;
 	}
+
+	/**
+	 * Gets payment method object from request.
+	 *
+	 * @throws PaymentRequestException
+	 */
+	public function get_payment_method(): PaymentMethodInterface {
+		$request = $this->request;
+
+		$is_woopayment_selected = isset( $request['payment_method'] ) && WC_Payment_Gateway_WCPay::GATEWAY_ID === $request['payment_method'];
+		if ( ! $is_woopayment_selected ) {
+			throw new PaymentRequestException( __( 'WooPayments is not used during checkout.', 'woocommerce-payments' ) );
+		}
+
+		if ( ! empty( $request['wcpay-payment-method'] ) ) {
```

Comment:
Let's check whether the key exists first, to allow triggering notices.

More importantly, I think the logic here is a bit flawed: Even though the hidden field named `wcpay-payment-method` contains the tokenized payment method, and is only created/populated after the submission of the form, it is possible that the form could be re-submitted (after an error) with a new payment method entered, but a saved one selected. 

If you take a look at the existing `Payment_Information` class, generally both the new payment method is loaded (if available), and the selected token. However, the token takes precedence:

https://github.com/Automattic/woocommerce-payments/blob/1a9176f8104328aafc73c2f94ff06ca2d78d5155/includes/class-payment-information.php#L166-L172

Reply:
> Let's check whether the key exists first, to allow triggering notices.

I think `empty` actually does both: check if the key exists, and ensure it's not an empty string `''`. https://3v4l.org/ugCT2

> More importantly, I think the logic here is a bit flawed:

This is very true! I fixed it here 74b0bc4b4a8892f0953eb0f6f3f1688f692d398a 
Actually, value of `wc-woocommerce_payments-payment-token` is `new`, and then `wcpay-payment-method` is attached when a new payment is added,

Reply:
> I think empty actually does both: check if the key exists, and ensure it's not an empty string ''. https://3v4l.org/ugCT2

Today I learned! https://www.php.net/manual/en/function.empty.php:

> empty() does not generate a warning if the variable does not exist.